kube-runtime/src/utils/stream_subscribe.rs

```diff
@@ -0,0 +1,243 @@
+
+/// Exposes the [`StreamSubscribe::subscribe()`] method which allows additional
+/// consumers of events from a stream without consuming the stream itself.
+///
+/// If a subscriber begins to lag behind the stream, it will receive an [`Error::Lagged`]
+/// error. The subscriber can then decide to abort its task or tolerate the lost events.
+///
+/// If the [`Stream`] is dropped or ends, any [`StreamSubscribe::subscribe()`] streams
+/// will also end.
+///
+/// ## Warning
+///
+/// If the primary [`Stream`] is not polled, the [`StreamSubscribe::subscribe()`] streams
+/// will never receive any events.
+#[pin_project]
+#[must_use = "subscribers will not get events unless this stream is polled"]
+pub struct StreamSubscribe<S>
+where
+    S: Stream,
+{
+    #[pin]
+    stream: S,
+    sender: broadcast::Sender<Option<Arc<S::Item>>>,
+}
+
+impl<S: Stream> StreamSubscribe<S> {
+    pub fn new(stream: S) -> Self {
+        let (sender, _) = broadcast::channel(CHANNEL_CAPACITY);
+
+        Self { stream, sender }
+    }
+
+    /// Subscribe to events from this stream
+    #[must_use = "streams do nothing unless polled"]
+    pub fn subscribe(&self) -> impl Stream<Item = Result<Arc<S::Item>, Error>> {
+        stream::unfold(self.sender.subscribe(), |mut rx| async {
+            match rx.recv().await {
+                Ok(Some(obj)) => Some((Ok(obj), rx)),
+                Err(RecvError::Lagged(amt)) => Some((Err(Error::Lagged(amt)), rx)),
+                _ => None,
+            }
+        })
+    }
+}
+
+impl<S: Stream> Stream for StreamSubscribe<S> {
+    type Item = Arc<S::Item>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.project();
+        let item = this.stream.poll_next(cx);
+
+        match item {
+            Poll::Ready(Some(item)) => {
+                let item = Arc::new(item);
+                this.sender.send(Some(item.clone())).ok();
+                Poll::Ready(Some(item))
+            }
+            Poll::Ready(None) => {
+                this.sender.send(None).ok();
+                Poll::Ready(None)
+            }
+            Poll::Pending => Poll::Pending,
+        }
+    }
+}
+
+/// An error returned from the inner stream of a [`StreamSubscribe`].
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub enum Error {
+    /// The subscriber lagged too far behind. Polling again will return
+    /// the oldest event still retained.
+    ///
+    /// Includes the number of skipped events.
+    Lagged(u64),
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Error::Lagged(amt) => write!(f, "subscriber lagged by {amt}"),
+        }
+    }
+}
+
+impl std::error::Error for Error {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use futures::{pin_mut, poll, stream, StreamExt};
+
+    #[tokio::test]
+    async fn stream_subscribe_continues_to_propagate_values() {
```

Comment:
hm, how are you running these tests?
i was on your branch trying `cargo test -p kube-runtime --features=unstable-runtime-subscribe --lib` and i'm not seeing any `subscribe` results, not even with `--all-features` makes me think that some of the feature selection stuff is not working correctly

they don't show up in CI unit tests either: https://github.com/kube-rs/kube/actions/runs/4205152176/jobs/7296822788 but that's probably just a missing `--all-features` to add to https://github.com/kube-rs/kube/blob/main/.github/workflows/ci.yml#L46-L47 once we figure out why they don't show up at all)